### DISCUSSION TITLE
Proposal: ensure read-only consistency of parameters between all primary constructor supporting types

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#7248] [Language Ideas] [answered]
For records, the read-only behaviour of primary constructor parameters is as follows: 

1. The parameters are mutable during construction and initialisation.
2. For record classes and readonly record structs, the parameters become immutable outside of that construction and initialisation stage
3. For record structs, remain mutable outside of that construction and initialisation stage

```cs
public record class RC1(int x, int y) 
{
    public int P
    {
        get => x;
        init => x = value;
    }
    
    public int Total() => x++ + y; // error
}

public record struct RS1(int x, int y) 
{
    public int P
    {
        get => x;
        init => x = value;
    }
    
    public int Total() => x++ + y; // all good
}


public readonly record struct RS2(int x, int y) 
{
    public int P
    {
        get => x;
        init => x = value;
    }
    
    public int Total() => x++ + y; // error 
}
```

This behaviour is well established and understood. Therefore there will likely be an expectation that primary constructor parameters for non-record types will follow the same rules. This happens for structs and readonly structs, but classes behave differently to what's expected:
```cs

public struct S1(int x, int y) 
{
    public int P
    {
        get => x;
        init => x = value;
    }
    
    public int Total() => x++ + y; // All good, as expected
}

public readonly struct S2(int x, int y) 
{
    public int P
    {
        get => x;
        init => x = value;
    }
    
    public int Total() => x++ + y; // Error, as expected
}

public class C1(int x, int y) 
{
    public int P
    {
        get => x;
        init => x = value;
    }
    
    public int Total() => x++ + y; // An error iks expected here but there isn't one
}
```

This proposal is that this inconsistency be addressed. Clearly allowing the parameter to become mutable is record classes is undesirable. Therefore the proposed solution is that classes change to behave like record classes before the non-record primary constructor feature is released.

**Comment by HaloFour** [comment]
I don't think framing it as an inconsistency is accurate.  What readonly-ness occurs in records is the result of the opinionated member synthesis, specifically that records emit `init`-only properties.  Non-record classes do not do this, nor do they synthesize storage unless necessary.

I can understand the argument of wanting primary constructor arguments readonly, but it runs contrary to the rest of the language.  In many other languages that have primary constructors, like Scala and Kotlin, the parameters aren't implicitly readonly, but gain that trait through declaring the parameter readonly.  Java is an outlier, but that's because primary constructors are linked to records, which are much more opinionated than they are in C#.

**Reply by DavidArno** [reply]
> I can understand the argument of wanting primary constructor arguments readonly, but it runs contrary to the rest of the language.

Primary constructor parameters behave in a read-only fashion for record classes, readonly record structs and readonly structs. I do not understand the reasoning behind you wanting primary constructor parameters for classes to fit it with "the rest of the language" (ie, methods) rather than with the behaviour of other primary constructor parameters.

**Reply by ufcpp** [reply]
Primary constructor parameters is not readonly.

```cs
Console.WriteLine(new X(0));

readonly record struct X(int A)
{
    public int A { get; } = A = 1;
    public int B { get; } = A;
}
```

**Reply by CyrusNajmabadi** [reply]
> Primary constructor parameters behave in a read-only fashion for record classes,

This is not the case. From the language's perspective, for a record, you are mostly not accessing or manipulating the parameter. You are accessing or manipulating a property/field (the rules around which depend on what sort of record you have).  

In *all* cases, if you actually work with the parameter, it is not readonly itself.  PCs for classes/structs behave the same way for parameters. 

For example, with record classes, you are commonly interacting with an init property. You can observe this given that says property can be made virtual and then overridden. If you do access the parameter though, you can write to it, and that is observable. 

We have some opinionated defaults for the synthesized members for records. But they're just about records (which are opinionated on everything, that's the point), and only about the members, not the parameters. 

Normal classes/structs are the non-opinionated, flexible, system on the opposite end if the spectrum from records. I do not expect (or even think it's appropriate) for classes to pull in the opinions of records back into them. For example, it is very intentional that you do not get member synthesis with classes+PC.  On the class/struct side, a PC is just a constructor, and parameters are just parameters. 

Note: we may introduce features to let classes/structs opt into synthesis. But it would be an orthogonal, augmenting, feature.  :-)

**Reply by DavidArno** [reply]
@ufcpp
> Primary constructor parameters is not readonly.

As I clearly state in the details of current PC parameter behaviour, PC parameters are mutable for all types in the construction/initialisation phase. This proposal relates to the behaviour outside of that construction/initialisation phase.

**Reply by CyrusNajmabadi** [reply]
There aren't any 'phases' from teh perspective of the language.  Any place where you are referring to the parameter, it works like any other parameter would (and thus is 'writable' as we have no 'readonly' concept for parameters yet).  If you find yourself in a situation where you cannot write to something, that's because you are not referring to the parameter, but to some real member of the type, and it is the member which defines (and can change) what its read/write semantics are.

**Reply by DavidArno** [reply]
> There aren't any 'phases' from teh perspective of the language...

`error CS9114: A primary constructor parameter of a readonly type cannot be assigned to (except in init-only setter of the type or a variable initializer)`

Your move...

**Reply by CyrusNajmabadi** [reply]
That isn't anything due to 'phases'.  That's what `readonly` means.  In other words, we have a `readonly` "concept" that can be applied to a type.  That concept goes through and augments a bunch of rules to describe what impact it has (specifically in terms of restricting capabilities).  You can see that impact for many different features.  The above part is then how it impacts *this* feature we're discussing (PC parameters).

My explanation refers to PC parameters in the absence of that obvious contextual limitation.  "readonlyness" is not a 'phase'.  it's a set of restrictions (and allowed optimizations due to those restrictions).

**Comment by DavidArno** [comment, marked as answer]
Whilst there appears an inconsistency in the behaviour of PC parameters between different types, this apparent inconsistency is just an accidental side-effect of the behaviour of the compiler. During the lowering process, the compiler will - under certain conditions - synthesise a member with the same name of the parameter. The compiler then "hides" the parameter and uses the member in subsequent references to that token. If the member has read-only behaviour, the parameter appears to become read-only.

As the inconsistency isn't real and is instead a side-effect of compiler behaviour, there appears no compelling desire to address that inconsistent behaviour. So I'm closing this proposal.

**Reply by CyrusNajmabadi** [reply]
> just an accidental side-effect of the behaviour of the compiler.

No.  It's was very intentional, and part of the *language* specification on how things would work.

> During the lowering process, the compiler will - under certain conditions - synthesise a member with the same name of the parameter.

This is not the case (for Roslyn) for normal PC parameters.  We actually do not synthesize any members with the same name as the parameter.  A compliant impl would be allowed to do this, but we would recommend against it (as you could easily end up with duplicate names in metadata).  In the Roslyn case, we have chose to encode this as fields with *different* names.  But that's an impl detail.  Other options that would still be viable (and the compielr might use in the future, or in certain circumstances) could be:

1. emitting a different type, and storing the information within an instance of that wrapper type.
2. using something like a ConditionalWeakTable to associate the data in another location entirely.

--

Now, for *records*, we do generate a member.  This is not an "accidental side effect".  The is literally how the language is spec'ed to behave, and all compliant implementations of the language would have to follow that behavior.  

> The compiler then "hides" the parameter and uses the member in subsequent references to that token

This is an ok statement depending on how you interpret things.  What really happens is that the members are simply in scope in a closer context at certain lookup points.  Nothing is explicitly "hidden/removed/etc.".  But you could call the aspect of scoping a way to indicate something is 'hidden'.  Note that there is a subtle, but important, distinction here.  Specifically, "hiding" would mean that the parameter would be gone, and you would never be able to use it.  Whereas, if there's something closer in scope, that means you coudl potentially use the thing higher in scope *if* the language provides a means to get to it.  For example, a local-variable doesn't "hide" a same-named member in the type.  it is simply closer it scope.  But hte member is still totally available through things like `this.`.   

> and uses the member in subsequent references to that token. If the member has read-only behaviour, the parameter appears to become read-only.

Nothing changes about the parameter.  For example, you could have:

```c#
public record C(int X) {
    public int Y { get; set; } = X;
}
```

Even though the compiler will have synthesized an 'X' property, it doesn't use that member "in subsequent references to that token".  The 'X' in the 'Y' property still refers to the 'X" parameter, not the property.  This is very relevant (As mentioned in an earlier response) as things like properties could be virtual.  And, as such, you might then be able to observe differences by calling through teh virtual property vs using the parameter. 

Instead, what is more correct to say is that both the parameters and members exist, and scoping determines which you see at a particular location in the rest of teh code. 

And, if we ever add a feature to look up a local/parameter that may be in a higher scope, you would now be able to use these in those particular locations, even though a synthesized (or non-synthesized) member existed with the same name.  This is similar to how `::` (and `this.` etc.) exists to ensure one can reference certain names, even if there are intermediary names that would otherwise be seen first.

**Reply by CyrusNajmabadi** [reply]
> As the inconsistency isn't real and is instead a side-effect of compiler behaviour,

This is not 'compiler behavior'.  This is the definition of the feature as per the language :)
